lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -447,6 +455,9 @@ public void OnNext(ICompletedTask completedTask)
                         _taskManager.RecordCompletedTask(completedTask);
                         TryRecovery();
                         break;
+                    case SystemState.Fail:
```

Comment:
Are you sure it's unexpected ? as I understand the events are asynchronous. The cancelled event might trigger Fail state, while there are still events from other nodes.